---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# CNVds

<!-- badges: start -->
<!-- badges: end -->

The goal of CNVds is to ...

## Description

## Installation

You can install the development version of CNVds from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("jenydu/CNVds")
```

## Overview

## Contributions

## References

Alexander-Bloch, A. et al. Copy number variant risk scores associated with cognition, psychopathology, and brain structure in youths in the philadelphia neurodevelopmental cohort. JAMA Psychiatry 79, 699–709 (2022).

Collins, R. L. et al. A cross-disorder dosage sensitivity map of the human genome. Cell 185, 3041-3055.e25 (2022).

Karczewski, K. J. et al. The mutational constraint spectrum quantified from variation in 141,456 humans. Nature 581, 434–443 (2020).

Ming, C, Wang, M, Wang, Q, et al. Whole genome sequencing–based copy
number variations reveal novel pathways and targets in Alzheimer's disease.
\emph{Alzheimer's & Dementia} 18, 1846- 1867 (2022).
\href{https://doi.org/10.1002/alz.12507}{Link}

## Acknowledgements

This package was developed as part of an assessment for 2019-2022 BCB410H: Applied Bioinformatics course at the University of Toronto, Toronto, CANADA. `CNVds` welcomes issues, enhancement requests, and other contributions. To submit an issue, use the [GitHub issues](https://github.com/jenydu/CNVds/issues). Many thanks to those who provided feedback to improve this package.
