---
title: "Introduction_CNVds"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction_CNVds}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
# The following codes are adapted from Testing package: Introduction_TestPackage.Rmd
# Author: Anjali Silva
library(knitr)
opts_chunk$set(fig.align = "center", 
               out.width = "90%",
               fig.width = 6, fig.height = 5.5,
               dev.args=list(pointsize=10),
               par = TRUE, # needed for setting hook 
               collapse = TRUE, # collapse input & ouput code in chunks
               warning = FALSE)
knit_hooks$set(par = function(before, options, envir)
  { if(before && options$fig.show != "none") 
       par(family = "sans", mar=c(4.1,4.1,1.1,1.1), mgp=c(3,1,0), tcl=-0.5)
})
set.seed(1122) # for exact reproducibility
```

```{r setup1}
library(CNVds)
```

## Introduction

'CNVds' is an R package developed to evaluate the pathogenicities of human CNVs based on the dosage sensitivity scores (probability of loss intolerance, probability of haploinsufficiency, and probability of triplosensitivity) of the genes that are encompassed within. This package is developed for BCB410H: Applied Bioinformatics at the University of Toronto, Canada. **This document gives a tour of CNVds (version 0.1.0)**. It was written in R Markdown, using the [knitr](https://cran.r-project.org/package=knitr) package for production. See `help(package = "CNVds")` for further details.


To download **CNVds**, use the following commands:

``` r
require("devtools")
install_github("jenydu/CNVds", build_vignettes = TRUE)
library("CNVds")
```

To list all sample functions available in the package:

``` r
ls("package:CNVds")
```

To list all sample datasets available in the package:

``` r
data(package = "CNVds")
```

<br>

## Components

<div style="text-align:center"><img src="DU_J_A1.png" alt="DU_J_A1" width="500"/>

<div style="text-align:left">

### Minimal requirements

-   The **R subdirectory** within an R package contains all the R code files or script files. These are files ending with ".R".

-   The **man subdirectory** contains help documentation for each function that is available to the user. Help files should be autogenerated via *roxygen2* package based on the roxygen tags used in the .R files from R subdirectory.

-   **DESCRIPTION** file contains metadata, including the package name, the version number, the author and maintainer contact information, along with other information.

-   **NAMESPACE** file provides a "context for looking up the value of an object associated with a name". This file will contain export(), import() and importFrom() statements. This file will be auto-updated via *roxygen2* package based on the roxygen tags used in the .R files from R subdirectory.

### Other

-   **README** file mainly describes what the package does. For the purpose of this course, the following elements in order are required: 1. Description, 2. Installation, 3. Overview, 4. Contributions, 5. References, and 6. Acknowledgements. Any other content of student's choice should be placed after these six elements.

-   **LICENSE** file describing who can use the R package.

-   The **data** subdirectory with data (.rda) is included to demonstrate utility of functions in the package with this data. If data is included, proper documentation for data must be provided using data.R file. Including data in the R package, however, is optional. If not providing own data, must show examples with external datasets so user of the package can still understand utility of each function.

-   The **tests** subdirectory includes the testthat subdirectory, which has test that are written to see if functions are working as expected. There should be unit tests and integration tests.

-   The **vignettes** subdirectory permit to include vignettes that provide an overview of the R package.

-   The **inst** subdirectory contains other files and subdirectories. **CITATION** file should be included here. Other subdirectories may include extdata and shiny-scripts. The **extdata** subdirectory can include images included in README file. The **shiny-scripts** subdirectory should include shiny script files (app.R).

-   The local project must be housed in a GitHub repository. There should be a record of the changes made to the package via Git commit messages. For the purpose of this course, ensure the **.Rproj** file is available on GitHub. Do NOT add it to .gitignore.

<br>

<div style="text-align:left">

## Package References

[Silva, A. (2020) TestingPackage: An Example R Package For
  BCB410H. Unpublished.](https://github.com/anjalisilva/TestingPackage)

<br>

## References

Alexander-Bloch, A. et al. (2022). Copy number variant risk scores associated with cognition, psychopathology, and brain structure in youths in the philadelphia neurodevelopmental cohort. *JAMA Psychiatry* 79, 699–709. [https://doi.org/10.1001/jamapsychiatry.2022.1017](https://doi.org/10.1001/jamapsychiatry.2022.1017)

Collins, R. L. et al. (2022). A cross-disorder dosage sensitivity map of the human genome. *Cell* 185, 3041-3055.e25. [https://doi.org/10.1016/j.cell.2022.06.036](https://doi.org/10.1016/j.cell.2022.06.036)

Jgraph. (2022). Drawio-desktop. Image created by Du, J. Retrieved November 14, 2022, from [https://app.diagrams.net/](https://app.diagrams.net/)

Karczewski, K. J. et al. (2020). The mutational constraint spectrum quantified from variation in 141,456 humans. *Nature* 581, 434–443. [https://doi.org/10.1038/s41586-020-2308-7](https://doi.org/10.1038/s41586-020-2308-7)

Ming, C, Wang, M, Wang, Q, et al. (2022). Whole genome sequencing–based copy
number variations reveal novel pathways and targets in Alzheimer's disease.
*Alzheimer's & Dementia* 18, 1846-1867. [https://doi.org/10.1002/alz.12507](https://doi.org/10.1002/alz.12507)

R Core Team (2013). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. ISBN 3-900051-07-0. [http://www.R-project.org/](http://www.R-project.org/)

Silva, Anjali. (2022). Anjalisilva/TestingPackage: A Simple R Package Illustrating Components of an R Package: 2019-2022 BCB410H - Applied Bioinformatics, University of Toronto, Canada. GitHub. [https://github.com/anjalisilva/TestingPackage](https://github.com/anjalisilva/TestingPackage)

Wickham H (2016). ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York. ISBN 978-3-319-24277-4. [https://ggplot2.tidyverse.org](https://ggplot2.tidyverse.org)

Wickham H, Girlich M (2022). tidyr: Tidy Messy Data. [https://tidyr.tidyverse.org](https://tidyr.tidyverse.org)

------------------------------------------------------------------------

```{r}
sessionInfo()
```
