---
title: "Introduction_CNVds"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction_CNVds}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
# The following codes are adapted from Testing package: Introduction_TestPackage.Rmd
# Author: Anjali Silva
library(knitr)
opts_chunk$set(fig.align = "center", 
               out.width = "90%",
               fig.width = 6, fig.height = 5.5,
               dev.args=list(pointsize=10),
               par = TRUE, # needed for setting hook 
               collapse = TRUE, # collapse input & ouput code in chunks
               warning = FALSE)
knit_hooks$set(par = function(before, options, envir)
  { if(before && options$fig.show != "none") 
       par(family = "sans", mar=c(4.1,4.1,1.1,1.1), mgp=c(3,1,0), tcl=-0.5)
})
set.seed(1122) # for exact reproducibility
```

```{r setup1}
library(CNVds)
```

## Introduction

'CNVds' is an R package developed to evaluate the pathogenicities of human CNVs based on the dosage sensitivity scores (probability of loss intolerance, probability of haploinsufficiency, and probability of triplosensitivity) of the genes that are encompassed within. This package is developed for BCB410H: Applied Bioinformatics at the University of Toronto, Canada. **This document gives a tour of CNVds (version 0.1.0)**. It was written in R Markdown, using the [knitr](https://cran.r-project.org/package=knitr) package for production. See `help(package = "CNVds")` for further details.

To download **CNVds**, use the following commands:

``` r
require("devtools")
install_github("jenydu/CNVds", build_vignettes = TRUE)
library("CNVds")
```

To list all sample functions available in the package:

``` r
ls("package:CNVds")
```

To list all sample datasets available in the package:

``` r
data(package = "CNVds")
```

<br>

## Components

<img src="DU_J_A1.png" alt="DU_J_A1" width="800"/>

## Sample Usage
Annotate the CNV regions from sampleInputCNV.rda, and plot both the CNV region
distribution and pHI score distribution.

```
sampleInputCNV <- NULL
load('~/CNVds/data/sampleInputCNV.rda')
plotCNVcounts(sampleInputCNV[c(1:4)])

annotated <- NULL
for (i in seq_along(1:nrow(sampleInputCNV))) {
  output <- annotateCNV(sampleInputCNV[i, 1], sampleInputCNV[i, 2],                       
                        sampleInputCNV[i, 3], sampleInputCNV[i, 4],
                        sampleInputCNV[i, 5], reference = 'GRCh37')
  annotated <- rbind(annotated, output)
  }

  pHIscores <- findpHI(annotated$gene)
  annotated <- merge(annotated, pHIscores, by='gene')
  plotScoresByChr(annotated[c('chr', 'gene', 'pHI')], 'pHI', 0.8)
}
```

## Package References

[Silva, A. (2020) TestingPackage: An Example R Package For BCB410H. Unpublished.](https://github.com/anjalisilva/TestingPackage)

<br>

## References

Alexander-Bloch, A. et al. (2022). Copy number variant risk scores associated with cognition, psychopathology, and brain structure in youths in the philadelphia neurodevelopmental cohort. *JAMA Psychiatry* 79, 699--709. <https://doi.org/10.1001/jamapsychiatry.2022.1017>

Collins, R. L. et al. (2022). A cross-disorder dosage sensitivity map of the human genome. *Cell* 185, 3041-3055.e25. <https://doi.org/10.1016/j.cell.2022.06.036>

Jgraph. (2022). Drawio-desktop. Image created by Du, J. Retrieved November 14, 2022, from <https://app.diagrams.net/>

Karczewski, K. J. et al. (2020). The mutational constraint spectrum quantified from variation in 141,456 humans. *Nature* 581, 434--443. <https://doi.org/10.1038/s41586-020-2308-7>

Ming, C, Wang, M, Wang, Q, et al. (2022). Whole genome sequencing--based copy number variations reveal novel pathways and targets in Alzheimer's disease. *Alzheimer's & Dementia* 18, 1846-1867. <https://doi.org/10.1002/alz.12507>

R Core Team (2013). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. ISBN 3-900051-07-0. <http://www.R-project.org/>

Silva, Anjali. (2022). Anjalisilva/TestingPackage: A Simple R Package Illustrating Components of an R Package: 2019-2022 BCB410H - Applied Bioinformatics, University of Toronto, Canada. GitHub. <https://github.com/anjalisilva/TestingPackage>

Wickham H (2016). ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York. ISBN 978-3-319-24277-4. <https://ggplot2.tidyverse.org>

Wickham H, Girlich M (2022). tidyr: Tidy Messy Data. <https://tidyr.tidyverse.org>

------------------------------------------------------------------------

```{r}
sessionInfo()
```
